packages/request/lib/Agent.js

```diff
@@ -0,0 +1,341 @@
+//
+// SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const EventEmitter = require('events')
+const http2 = require('http2')
+const net = require('net')
+const crypto = require('crypto')
+const { promisify } = require('util')
+const QuickLRU = require('quick-lru')
+const { globalLogger: logger } = require('@gardener-dashboard/logger')
+const Semaphore = require('./Semaphore')
+
+const kSemaphore = Symbol('semaphore')
+const kRequest = Symbol('request')
+const kSid = Symbol('sid')
+const kTimestamp = Symbol('timestamp')
+const kConnected = Symbol('connected')
+const kTimeoutId = Symbol('timeoutId')
+const kIntervalId = Symbol('intervalId')
+
+const { NGHTTP2_CANCEL } = http2.constants
+
+class Agent extends EventEmitter {
+  constructor (options = {}) {
+    super()
+    const {
+      maxCachedTlsSessions = 100,
+      timeout = 60000,
+      connectTimeout = 5000,
+      gracePeriod = 5000,
+      maxOutstandingPings = 2,
+      pingInterval = 30000
+    } = options
+    this.timeout = timeout
+    this.connectTimeout = connectTimeout
+    this.gracePeriod = gracePeriod
+    this.maxOutstandingPings = maxOutstandingPings
+    this.pingInterval = pingInterval
+    this.tlsSessionCache = new QuickLRU({
+      maxSize: maxCachedTlsSessions
+    })
+    this.sessionMap = new Map()
+  }
+
+  destroy () {
+    for (const iterable of this.sessionMap.values()) {
+      for (const session of iterable) {
+        this.clearSessionTimers(session)
+        this.deleteSession(session, { force: true })
+        session.destroy()
+        this.tlsSessionCache.delete(session[kSid])
+      }
+    }
+  }
+
+  findSession (sid) {
+    const iterable = this.sessionMap.get(sid)
+    if (iterable) {
+      return Array.from(iterable)
+        // consider free sessions only
+        .filter(({ [kSemaphore]: semaphore }) => {
+          return semaphore.value > 0
+        })
+        // session with the highest load first
+        .sort(({ [kSemaphore]: a }, { [kSemaphore]: b }) => {
+          return a.value - b.value
+        })
+        .shift()
+    }
+  }
+
+  getSession (authority, { settings, ...options } = {}) {
+    authority = normalizeAuthority(authority)
+    options = Object.assign({
+      peerMaxConcurrentStreams: 100,
+      settings: Object.assign({
+        enablePush: false
+      }, settings),
+      connectTimeout: this.connectTimeout,
+      timeout: this.timeout,
+      maxOutstandingPings: this.maxOutstandingPings,
+      pingInterval: this.pingInterval
+    }, options)
+    const sid = getSessionId(authority, options)
+    let session = this.findSession(sid)
+    if (!session) {
+      session = this.constructor.createSession(authority, options)
```

Comment:
```suggestion
      if (this.tlsSessionCache.has(sid)) {
        options.session = this.tlsSessionCache.get(sid)
      }
      session = this.constructor.createSession(authority, options)
```
Check if the tlsSession can be invalid or outdated. Or the server has restartred. Include an ende 2 end test.

Reply:
✔️

Reply:
Invalid tls session tickets are ignored and a new valid one is generated and returned.